WL-4296 centralise TII configuration

bbbbgarcia opened the pull request:
TII account settings will always be retrieved from the global LTI tool. Endpoint url setting is still needed.

bbbbgarcia committed:
WL-4296 centralise TII configuration

buckett commented on contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java:
```diff
@@ -73,17 +74,21 @@ public void init() {
-		}			
-		endpoint = serverConfigurationService.getString("turnitin.ltiURL", "https://sandbox.turnitin.com/api/lti/1p0/");		
 		turnitinSite = serverConfigurationService.getString("turnitin.lti.site", "!turnitin");
+		if(StringUtils.isEmpty(turnitinSite)){
```
You supply a default in your call to the `ServerConfigurationService` so it should always be defined. The only way this would ever fire is if someone explicitly set `turnitin.lti.site=` which seems to be quite a edge case.

buckett replied:
Doesn't need removing, was just a comment.

buckett commented on contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java:
```diff
@@ -172,23 +176,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
+	public boolean obtainGlobalTurnitinLTIToolData(){
+		log.debug("Setting global TII LTI tool properties");
+		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
+		if ( tools == null || tools.size() < 1  || tools.size() > 1) {
```
Isn't it cleaner todo `tools == null || tools.size() != 1` .

buckett commented on contentreview-impl/tool/pom.xml:
```diff
@@ -41,6 +41,10 @@
             <groupId>org.sakaiproject.contentreview</groupId>
             <artifactId>contentreview-service-api</artifactId>
         </dependency>
+		<dependency>
```
It's considered bad practice to include the `impl` in the tool as you should be going through the API. If there are utility classes they should not be in the `impl`

bgarciaentornos replied:
I know, like I said elsewhere didn't want to include a such specific TII LTI method (obtainGlobalToolData) on the CR-API. Like this tool only has the servlets I added doesn't seem like a problem.

buckett commented on contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/GradingCallbackServlet.java:
```diff
@@ -196,6 +208,21 @@ protected void doPostXml(HttpServletRequest request, HttpServletResponse respons
         return;
     }
 	
+	public Map<String,Object> obtainGlobalTurnitinLTITool(String turnitinSite){
```
Wouldn't it be better to expose the LTI configuration from the service and just have the grading callback servlet get the config from there, rather than duplicating the code?

bgarciaentornos replied:
Didn't want to create a "obtainGlobalTurnitinLTITool" method into the CR-api. I guess I could have extracted it to a different place, though.

buckett commented on contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/SubmissionCallbackServlet.java:
```diff
@@ -144,6 +155,21 @@ protected void doPostJSON(HttpServletRequest request, HttpServletResponse respon
     }
 	
+	public Map<String,Object> obtainGlobalTurnitinLTITool(String turnitinSite){
+		M_log.debug("Getting global TII LTI tool");
```
Again, duplicating the code can easily lead to fixes only being done in some places down the line.

bgarciaentornos replied:
Besides what I've said on the other comments, there was also a problem injecting/accessing the LTIService from a method outside the servlet.

buckett commented on contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/TIICallbackServlet.java:
```diff
@@ -123,6 +135,21 @@ protected void doPostJSON(HttpServletRequest request, HttpServletResponse respon
     }
 	
+	public Map<String,Object> obtainGlobalTurnitinLTITool(String turnitinSite){
+		M_log.debug("Getting global TII LTI tool");
```
Again, duplicated code.

bgarciaentornos commented:
Made the changes (extracting replicated methods to a utils class and some more) and squashed the new commit

buckett commented on contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/TIICallbackServlet.java:
```diff
@@ -83,8 +89,13 @@ protected void doPostJSON(HttpServletRequest request, HttpServletResponse respon
+		String turnitinSite = ServerConfigurationService.getString("turnitin.lti.site", "!turnitin");
+		Map<String,Object> tiiData = ServletUtils.obtainGlobalTurnitinLTITool(turnitinSite);
+		if(tiiData == null){
+			M_log.error("Turnitin global LTI tool does not exist or properties are wrongly configured.");
```
This needs better handling as it will NPE a few lines down.

buckett commented on contentreview-impl/tool/pom.xml:
```diff
@@ -84,7 +84,11 @@
+		<dependency>
+            <groupId>org.sakaiproject.basiclti</groupId>
+            <artifactId>basiclti-api</artifactId>
+			<version>10-SNAPSHOT</version>
```
Indentation.

buckett commented on contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/ServletUtils.java:
```diff
@@ -0,0 +1,33 @@
+			} else {
+				M_log.warn("Found: " + tools.size());
+			}
+			M_log.error("obtainGlobalTurnitinLTIToolData: wrong global TII LTI tool configuration");
```
It's much better to combine the logging into one statement with the message from the previous warning. Otherwise if you have error messages going somewhere, but warn level not you won't see why it failed.

buckett commented on contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/GradingCallbackServlet.java:
```diff
@@ -98,8 +103,13 @@ protected void doPostXml(HttpServletRequest request, HttpServletResponse respons
+		String turnitinSite = ServerConfigurationService.getString("turnitin.lti.site", "!turnitin");
+		Map<String,Object> tiiData = ServletUtils.obtainGlobalTurnitinLTITool(turnitinSite);
+		if(tiiData == null){
+			M_log.error("Turnitin global LTI tool does not exist or properties are wrongly configured.");
```
If `tiiData` is null it's going to NPE a few lines down.

bgarciaentornos commented:
Updated the PR

buckett commented:
Still has outstanding comments.

bgarciaentornos commented:
Added return calls just after the tiiData errors (check the whole PR), if you refer to that. If you don't, please tell me where to look at.

buckett commented on contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java:
```diff
@@ -180,23 +184,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
+		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
+		if ( tools == null || tools.size() != 1 ) {
+			if(tools == null){
+				log.warn("No tools found");
```
Same logging issue.

buckett commented on contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java:
```diff
@@ -180,23 +184,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
-				return null;
+	public boolean obtainGlobalTurnitinLTIToolData(){
+		log.debug("Setting global TII LTI tool properties");
+		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
```
Doesn't this duplicate most of the code in ServletUtils?

bgarciaentornos replied:
This method is just for checking at the init method (returns a boolean),
the other one returns actual data for each servlet call.

2016-05-06 13:11 GMT+02:00 Matthew Buckett <notifications@github.com>:

> In
> contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java
> <https://github.com/ox-it/wl-turnitin/pull/17#discussion_r62316649>:
>
> > @@ -180,23 +184,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
> >  		return -5;
> >  	}
> >  	
> > -	public String getGlobalTurnitinLTIToolId(){
> > -		if(globalId == null){
> > -			log.debug("Setting global TII LTI tool id");
> > -			List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
> > -			if ( tools == null || tools.size() < 1  || tools.size() > 1) {
> > -				if(tools == null)
> > -					log.warn("No tools found");
> > -				else
> > -					log.warn("Found: " + tools.size());				
> > -				log.error("getGlobalTurnitinLTIToolId: wrong global TII LTI tool configuration");
> > -				return null;
> > +	public boolean obtainGlobalTurnitinLTIToolData(){
> > +		log.debug("Setting global TII LTI tool properties");
> > +		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
>
> Doesn't this duplicate most of the code in ServletUtils?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ox-it/wl-turnitin/pull/17/files/25e0348519e93241319d2ede100c98570a883e31#r62316649>
>

buckett replied:
But most of the code is the same.

bgarciaentornos replied:
Definitely. But we need it in both places (impl should not be included in
the tool).

2016-05-06 13:18 GMT+02:00 Matthew Buckett <notifications@github.com>:

> In
> contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java
> <https://github.com/ox-it/wl-turnitin/pull/17#discussion_r62317177>:
>
> > @@ -180,23 +184,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
> >  		return -5;
> >  	}
> >  	
> > -	public String getGlobalTurnitinLTIToolId(){
> > -		if(globalId == null){
> > -			log.debug("Setting global TII LTI tool id");
> > -			List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
> > -			if ( tools == null || tools.size() < 1  || tools.size() > 1) {
> > -				if(tools == null)
> > -					log.warn("No tools found");
> > -				else
> > -					log.warn("Found: " + tools.size());				
> > -				log.error("getGlobalTurnitinLTIToolId: wrong global TII LTI tool configuration");
> > -				return null;
> > +	public boolean obtainGlobalTurnitinLTIToolData(){
> > +		log.debug("Setting global TII LTI tool properties");
> > +		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
>
> But most of the code is the same.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ox-it/wl-turnitin/pull/17/files/25e0348519e93241319d2ede100c98570a883e31#r62317177>
>

buckett replied:
Why not move it to somewhere that both the tool and the impl and access?
Make it part of the API.

On 6 May 2016 at 12:23, bgarciaentornos <notifications@github.com> wrote:

> In
> contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java
> <https://github.com/ox-it/wl-turnitin/pull/17#discussion_r62317598>:
>
> > @@ -180,23 +184,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
> >  		return -5;
> >  	}
> >  	
> > -	public String getGlobalTurnitinLTIToolId(){
> > -		if(globalId == null){
> > -			log.debug("Setting global TII LTI tool id");
> > -			List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
> > -			if ( tools == null || tools.size() < 1  || tools.size() > 1) {
> > -				if(tools == null)
> > -					log.warn("No tools found");
> > -				else
> > -					log.warn("Found: " + tools.size());				
> > -				log.error("getGlobalTurnitinLTIToolId: wrong global TII LTI tool configuration");
> > -				return null;
> > +	public boolean obtainGlobalTurnitinLTIToolData(){
> > +		log.debug("Setting global TII LTI tool properties");
> > +		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
>
> Definitely. But we need it in both places (impl should not be included in
> the tool). 2016-05-06 13:18 GMT+02:00 Matthew Buckett <
> notifications@github.com>:
> … <#m_5664112745550056543_>
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/ox-it/wl-turnitin/pull/17/files/25e0348519e93241319d2ede100c98570a883e31#r62317598>
>



-- 
  Matthew Buckett, VLE Developer, IT Services, University of Oxford

buckett replied:
Were you going to fix this @bgarciaentornos?

bgarciaentornos replied:
I had a look at it and didn't find a quick way to put it all just on one
place, there were circular dependencies. It doesn't have anything to do
with the actual jira anyway.

2016-06-03 15:56 GMT+02:00 Matthew Buckett <notifications@github.com>:

> In
> contentreview-impl/impl/src/java/org/sakaiproject/turnitin/util/TurnitinLTIUtil.java
> <https://github.com/ox-it/wl-turnitin/pull/17#discussion_r65710754>:
>
> > @@ -180,23 +184,46 @@ public int makeLTIcall(int type, String urlParam, Map<String, String> ltiProps){
> >  		return -5;
> >  	}
> >  	
> > -	public String getGlobalTurnitinLTIToolId(){
> > -		if(globalId == null){
> > -			log.debug("Setting global TII LTI tool id");
> > -			List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
> > -			if ( tools == null || tools.size() < 1  || tools.size() > 1) {
> > -				if(tools == null)
> > -					log.warn("No tools found");
> > -				else
> > -					log.warn("Found: " + tools.size());				
> > -				log.error("getGlobalTurnitinLTIToolId: wrong global TII LTI tool configuration");
> > -				return null;
> > +	public boolean obtainGlobalTurnitinLTIToolData(){
> > +		log.debug("Setting global TII LTI tool properties");
> > +		List<Map<String, Object>> tools = ltiService.getToolsDao("lti_tools.site_id = '"+turnitinSite+"'", null, 0, 0, turnitinSite);
>
> Were you going to fix this @bgarciaentornos
> <https://github.com/bgarciaentornos>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ox-it/wl-turnitin/pull/17/files/25e0348519e93241319d2ede100c98570a883e31#r65710754>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AK6IluxK_9Rbd6N9QGjTMDGNFom0nidbks5qIDKHgaJpZM4H2DBM>
> .
>

buckett committed:
WL-4296 Improvements in centralising config.

Rather than duplicating code between both the tool and the impl we use a simple API to allow both parts to retrieve the values the same way.
